### ISSUE TITLE
about the implementation principle of Equal

### ISSUE THREAD

**Issue by TravisChenn1996** [type-challenges/type-challenges#8617]
Hi @antfu 关于 `utils/index.d.ts` 中的 `Equal` 方法，我认为有可读性更高的写法

```ts
// now
export type Equal<X, Y> =
  (<T>() => T extends X ? 1 : 2) extends
  (<T>() => T extends Y ? 1 : 2) ? true : false
```
```ts
// new
export type Equal<X, Y> = [X] extends [Y] ? true : false
```

**Comment by gomjellie** [comment]
@TravisChenn1996 

I think your solution should be like:
我觉得你的CODE应该是像这样

```ts
type Equal<T, U> = [T] extends [U] ? ([U] extends [T] ? true : false) : false
```

------

By the way have you understood how the original code works? can you explain?
I have no idea how T works
顺便说一下， 你可以说明 utils/index.d.ts 中的 Equal 方法
我不明了

```ts
type Equal<X, Y> =
  (<T>() => T extends X ? 1 : 2) extends
  (<T>() => T extends Y ? 1 : 2) ? true : false
```

**Comment by TravisChenn1996** [comment]
Hi @gomjellie，因为需要避免 `T` 或者 `U` 是 `any` 的情况
```ts
type Equal<X, Y> = (<T>() => T extends X ? 1 : 2) extends (<T>() => T extends Y ? 1 : 2) ? true : false
↓
type Equal<1, any> = (<any>() => any extends 1 ? 1 : 2) extends (<any>() => any extends any ? 1 : 2) ? true : false
↓
type Equal<1, any> = (1 | 2) extends (1) ? true : false
↓
false
```

**Comment by TravisChenn1996** [comment]
Hi @gomjellie, because of `Equal` need to avoid `X` or `Y` equals to `any`

```ts
type Equal<1, any>
↓
(<unknow>() => unknow extends 1 ? 1 : 2) extends (<unknow>() => unknow extends any ? 1 : 2) ? true : false
↓
(2) extends (1) ? true : false
```

**Comment by TravisChenn1996** [comment]
#9938

**Comment by jiangshanmeta** [comment]
@TravisChenn1996  As far as I am concerned, there are some corner cases that act different for two implementation.

For example:

```typescript
type A = Equal<{
  a:string;
  b:number
},{a:string}&{b:number}>

type B = Equal<{
  a:string;
},{
  readonly a:string
}>
```

Your implementation return true for both case, but the current implementation return false.

**Comment by LengSword** [comment]
@jiangshanmeta @PorkSashimi 
This seems to be because implementation of identity checks in typescript is not perfect.

[Here](https://stackoverflow.com/a/68963796) are some explanations for this.

TL;DR: `Equal<{x: 1} & {y: 2}, {x: 1, y: 2}>` is false, because typescript compares their `flags`. See [TypeFlags](https://github.com/microsoft/TypeScript/blob/0e905be42bd92d6679d43408deb818b09ff6c292/lib/tsserverlibrary.d.ts#L2547).
And the former has a flag `Intersection`, the latter does not. See [Intersection](https://github.com/microsoft/TypeScript/blob/0e905be42bd92d6679d43408deb818b09ff6c292/lib/tsserverlibrary.d.ts#L2517).

To solve this problem, you can do this:

```ts
type Compute<T> = { [K in keyof T]: T[K] }
type A = Equal<Compute<{ x: 1 } & { y: 2 }>, { x: 1, y: 2 }>  // should be true
```
